legend-engine-xt-relationalStore-executionPlan-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/state/ConnectionStateManager.java

```diff
@@ -356,13 +357,14 @@ public DataSourceWithStatistics getDataSourceForIdentityIfAbsentBuild(IdentitySt
         //Example: for kerberos based identities, creating a hikari pool is done as a PrivilegedAction which sets the subject for that security context
         //since each hikari pool run on its own thread, the pool implicit's subject is the creating one.
         //eventually kerberos credentials will expire(ie invalid) so we need to recreate the pool with the latest subject from the incoming request
-        if (!this.connectionPools.get(poolName).getIdentityState().isValid())
+        DataSourceWithStatistics dataSourceWithStatistics = this.connectionPools.get(poolName);
+        if (!dataSourceWithStatistics.getIdentityState().isValid() || (dataSourceWithStatistics.getStatistics().getLastConnectionRequestAge() > (evictionDurationInSeconds * 1000) && !dataSourceWithStatistics.hasActiveConnections()))
         {
             LOGGER.info("Pool [{}] for datasource [{}] does not have a valid identity state", principal, connectionKey.shortId());
             synchronized (poolLockManager.getLock(poolName))
             {
-                DataSourceWithStatistics dataSourceWithStatistics = this.connectionPools.get(poolName);
-                if (!dataSourceWithStatistics.getIdentityState().isValid())
+                dataSourceWithStatistics = this.connectionPools.get(poolName);
+                if (!dataSourceWithStatistics.getIdentityState().isValid() || (dataSourceWithStatistics.getStatistics().getLastConnectionRequestAge() > (evictionDurationInSeconds * 1000) && !dataSourceWithStatistics.hasActiveConnections()))
```

Comment:
as discussed this is not the right approach, you need to specify the expiry logic for the datasource so we can get rid of pools with expired or invalid credentials